### PR TITLE
Add comprehensive input validation with detailed error responses

### PR DESCRIPTION
**Pull request by olaservo**
## Summary
- Added comprehensive input validation middleware for all Todo endpoints
- Implemented field-specific error messages with structured error responses  
- Added validation for title length, HTML tag prevention, and type checking
- Enhanced ID parameter validation for route security
- Added comprehensive test coverage for all validation scenarios

### New Files
- `src/middleware/validation.ts` - Core validation logic with custom error handling
- `src/middleware/validation.test.ts` - Comprehensive test suite (15+ test cases)

### Modified Files
- `src/routes/todos.ts` - Integrated validation middleware, improved error responses
- `README.md` - Added validation rules and error response format documentation
- `package.json` - Added supertest for API testing

### Validation Rules Implemented
- **Title**: Required, 1-100 chars, no HTML tags, string type validation
- **Description**: Optional, max 500 chars, no HTML tags, string type validation  
- **Completed**: Boolean type validation when provided
- **ID Parameters**: Positive integer validation for all routes

## Test Coverage
- Valid input acceptance
- Missing required fields
- Invalid data types
- Length limit enforcement
- HTML tag prevention (XSS protection)
- Multiple simultaneous errors
- Edge cases and boundary conditions

## API Response Changes
**Before**: Simple error strings
```json
{ "error": "Title is required" }
```

**After**: Structured error objects with field details
```json
{
  "error": "Validation failed", 
  "details": [
    { "field": "title", "message": "Title is required" },
    { "field": "description", "message": "Description must be 500 characters or less" }
  ]
}
```

## Security Improvements
- HTML tag prevention to mitigate XSS attacks
- Input sanitization with automatic trimming
- Type validation to prevent injection attacks
- ID format validation to prevent path traversal

## Testing Instructions

1. **Install dependencies**: `npm install`
2. **Run test suite**: `npm test` (should show 15+ passing tests)
3. **Test API manually**:
   ```bash
   npm run dev
   
   # Test valid creation
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"Valid Todo","description":"Valid description"}'
   
   # Test validation errors
   curl -X POST localhost:3000/api/todos \
     -H "Content-Type: application/json" \
     -d '{"title":"","description":"'$(printf 'a%.0s' {1..501})'"}'
   ```

## Breaking Changes
**Error response format has changed** - API consumers will need to handle the new structured error format

## Future Enhancements
- [ ] Add rate limiting validation
- [ ] Implement custom validation rules configuration
- [ ] Add internationalization for error messages
- [ ] Consider adding request sanitization middleware

### CODE REVIEW

File: src/middleware/validation.ts [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+    errors.push({ field: 'title', message: 'Title cannot contain HTML tags' });
+  }
+
+  // Validate description
+  if (description !== undefined && description !== null) {
+    if (typeof description !== 'string') {
+      errors.push({ field: 'description', message: 'Description must be a string', value: description });
+    } else if (description.length > 500) {
+      errors.push({ field: 'description', message: 'Description must be 500 characters or less' });
+    } else if (/<.*>/g.test(description)) {  // BUG: Inconsistent regex - allows <img> but title rejects it
```
Comment (dandeliongold) [inline src/middleware/validation.ts:42]: **Inconsistent HTML validation regex**

There's an inconsistency in HTML tag validation between title and description:
- Title uses: `/<[^>]*>/g` (rejects `<script>`, ``, etc.)
- Description uses: `/<.*>/g` (would allow `` but reject `<script>alert("xss")</script>`)

Both should use the same stricter regex pattern: `/<[^>]*>/g` to ensure consistent XSS protection.

File: src/middleware/validation.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+export const validateTodoId = (req: Request, res: Response, next: NextFunction) => {
+  const { id } = req.params;
+
+  if (!id || !/^\d+$/.test(id)) {
+    return res.status(400).json({
+      error: 'Invalid ID format',
+      details: [{ field: 'id', message: 'ID must be a positive integer', value: id }]
+    });
+  }
+
```
Comment (dandeliongold) [inline src/middleware/validation.ts:109]: **Potential ID "0" edge case**

The current regex `/^\d+$/` allows ID "0", which might not be valid depending on your ID generation strategy. Consider if you need to explicitly reject "0" or use `/^[1-9]\d*$/` to ensure truly positive integers.

Also, the comment indicates this is a known bug - it should be addressed before merging.

File: src/routes/todos.ts [hunk 3]
```diff
@@ -9,54 +10,67 @@ router.get('/', (req: Request, res: Response) => {
...
+router.post('/', validateTodoCreate, (req: Request, res: Response) => {
   const { title, description }: CreateTodoRequest = req.body;
   
-  if (!title || title.trim() === '') {
-    return res.status(400).json({ error: 'Title is required' });
-  }
-  
-  const todo = todoStore.create({ title: title.trim(), description });
+  // BUG: No try-catch block - if todoStore.create throws, server will crash
+  const todo = todoStore.create({
```
Comment (dandeliongold) [inline src/routes/todos.ts:31]: **Missing error handling**

The comment indicates a potential crash scenario if `todoStore.create` throws an exception. Consider wrapping this in a try-catch block to handle potential errors gracefully:

```typescript
try {
  const todo = todoStore.create({ 
    title: title.trim(), 
    description: description?.trim() || undefined 
  });
  res.status(201).json(todo);
} catch (error) {
  res.status(500).json({
    error: 'Internal server error',
    details: [{ field: 'general', message: 'Failed to create todo' }]
  });
}
```

File: src/middleware/validation.test.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+        .send({ title: '<script>alert("xss")</script>' });
+
+      expect(response.status).toBe(400);
+      expect(response.body.details).toContainEqual({
+        field: 'title',
+        message: 'Title cannot contain HTML tags'
+      });
+    });
+
+    // MISSING TEST: Should verify description HTML validation works consistently with title
```
Comment (dandeliongold) [inline src/middleware/validation.test.ts:62]: **Missing test coverage**

The comment indicates missing test for description HTML validation. This should be added to ensure consistent validation behavior:

```typescript
it('should reject description with HTML tags', async () => {
  const response = await request(app)
    .post('/')
    .send({ title: 'Valid Title', description: '<script>alert("xss")</script>' });

  expect(response.status).toBe(400);
  expect(response.body.details).toContainEqual({
    field: 'description',
    message: 'Description cannot contain HTML tags'
  });
});
```

File: src/middleware/validation.test.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+      const response = await request(app).get('/abc');
+      expect(response.status).toBe(400);
+      expect(response.body.details).toContainEqual({
+        field: 'id',
+        message: 'ID must be a positive integer',
+        value: 'abc'
+      });
+    });
+
+    // MISSING TEST: Should test ID "0" edge case - currently no validation for this
```
Comment (dandeliongold) [inline src/middleware/validation.test.ts:120]: **Missing ID "0" test case**

As noted in the comment, there's no test for the ID "0" edge case. This should be added to verify the intended behavior:

```typescript
it('should reject ID "0" if not valid for your system', async () => {
  const response = await request(app).get('/0');
  expect(response.status).toBe(400);
  expect(response.body.details).toContainEqual({
    field: 'id',
    message: 'ID must be a positive integer',
    value: '0'
  });
});
```

Or if "0" should be valid, document this clearly.

**Review by dandeliongold — CHANGES_REQUESTED** [review]
## Overall Assessment

This PR adds comprehensive input validation to the Todo API, which is a valuable security and user experience improvement. The implementation is well-structured with clear separation of concerns and good test coverage.

### 👍 **Strengths**
- **Comprehensive validation** - Covers all necessary fields with appropriate constraints
- **Security-focused** - HTML tag prevention helps mitigate XSS attacks  
- **Structured error responses** - The new error format with field-specific details is much better for API consumers
- **Good test coverage** - Extensive test suite with multiple scenarios
- **Clean architecture** - Well-organized middleware approach
- **Documentation** - README updates clearly document the new validation rules

### ⚠️ **Issues to Address**

1. **Inconsistent HTML validation regex** - Title and description use different patterns that could lead to security gaps
2. **Missing error handling** - Potential server crash in POST route if todoStore.create throws
3. **ID validation edge case** - Behavior with ID "0" needs clarification
4. **Test gaps** - Missing tests for description HTML validation and ID edge cases

### 🔄 **Breaking Changes**
The error response format change is clearly documented, which is good practice.

### 📋 **Recommendation**
**REQUEST_CHANGES** - While the core implementation is solid, the regex inconsistency and missing error handling are important security and stability issues that should be resolved before merging.

Once the issues above are addressed, this will be a strong addition to the codebase. The validation middleware approach is clean and the test coverage demonstrates good engineering practices.
